src/main/java/hudson/plugins/build_timeout/BuildTimeoutWrapper.java

```diff
@@ -193,6 +195,8 @@ public synchronized void rescheduleIfScheduled() {
             public synchronized boolean tearDown(AbstractBuild build, BuildListener listener) throws IOException, InterruptedException {
                 if (task != null) {
                     task.cancel();
+                    // avoid memory leaks for the case where this timer is in the future (JENKINS-31627).
+                    Trigger.timer.purge();
```

Comment:
@tfennelly calling purge tells the timer to walk through the list of Tasks it has and remove any that are cancelled.

Its not as cheap as inserting new tasks - or letting them natrually drop of when they come to run - but the instance that showed this issue had ** over 3 million** of these objects - and there should be only a few hundred or so.  So whilst calling `purge()` can be expensive - it works out cheaper than letting them pile up and running out of memory.